### ISSUE TITLE
Extracting Gradient (and Hessian) from trees_to_dataframe

### ISSUE THREAD

**Issue by seanv507** [dmlc/xgboost#12319]
using xgboost 3.3.0 python interface

> I would like to reopen?
> 
> This is my understanding. 
> The 'Cover' column is the sum of Hessians
> For leaf nodes, the Gain column is the optimal weight $w^*_j$ equation 5 above. So one can recover the sum of gradients G, by multiplying by corresponding ('Cover' + lambda).
> For internal split nodes, the gain column is given by equation 7 in the paper (just below the screenshot pasted above)
> 
> Therefore, one should be able to recover the gradients by summing up the gradients of all the leaf nodes associated with an internal split node
> 
> 
> I cannot recover the Gain column by doing this.
> 
> Here is an example, with $\gamma=0$ and $\lambda=1$
> This is a dump from `clf.get_booster.trees_to_dataframe` using the python library.  I added the Hessian, Gradient,Obj  and Expected Gain columns. (formulas in excel sheet)
> 
> 
> <google-sheets-html-origin><style type="text/css"></style>
> Tree | Node | ID | Feature | Split | Yes | No | Missing | Gain | Cover | Category | Hessian | Gradient | Obj: Gradient^2/(Hessian+lambda) | Expected Gain |   | Lambda | 1
> -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
> 0 | 20 | 0-20 | Leaf |   |   |   |   | 0.335313559 | 2.33931756 |   | 2.33931756 | -1.119718456 | 0.3754567805 |   |   | Gamma | 0
> 0 | 19 | 0-19 | Leaf |   |   |   |   | 0.0298848841 | 1.40359056 |   | 1.40359056 | -0.07183102531 | 0.002146661866 |   |   |   |  
> 0 | 15 | 0-15 | f16 | 0.0313699991 | 0-19 | 0-20 | 0-20 | 0.869481564 | 3.74290824 |   | 3.74290812 | -1.191549481 | 0.2993501307 | 0.03912665581 |   |   |  
> 
> 
> 
> [Tree.xlsx](https://github.com/user-attachments/files/29840339/Tree.xlsx)
> 
> using xgboost scikitlearn interface demo code
> ```
> from sklearn.datasets import load_breast_cancer
> from sklearn.model_selection import train_test_split
> 
> import xgboost as xgb
> 
> X, y = load_breast_cancer(return_X_y=True)
> X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, random_state=94)
> 
> # Use "hist" for constructing the trees, with early stopping enabled.
> clf = xgb.XGBClassifier(tree_method="hist", early_stopping_rounds=2)
> # Fit the model, test sets are used for early stopping.
> clf.fit(X_train, y_train, eval_set=[(X_test, y_test)])
> ```
>  

 _Originally posted by @seanv507 in [#9048](https://github.com/dmlc/xgboost/issues/9048#issuecomment-4923458394)_

**Comment by trivialfis** [comment]
I wish I could design a new schema for the trees to dataframe ..

That said, the "Gain" for leaves is leaf weights. The weights are multiplied by the learning rate. As for internal split gain (which is the same column), it's multiplied by 2 for easier calculation (to remove the 1/2 factor in the middle of the gain function), which doesn't alter the ordering between split candidates.

**Comment by seanv507** [comment]
Great, those 2 changes explain the difference!
